### ISSUE TITLE
insertBy is not atomic

### ISSUE THREAD

**Issue by tim2CF** [yesodweb/persistent#1186]
I'm using postgres 11.

`insertBy` function is supposed to return `Left old` in case where row with given unique key already exists, or `Right new` in case where given new record has been inserted. But in some cases, where `insertBy` is applied from different threads - it just not working this way, it's throwing runtime exception instead, because `insert` part of `insertBy` always assumes that record does not exist, which might be not true in case of mutithreading where second thread did insert right after first thread checked row existence. The problem there is that `insertBy` is not atomic, it's sequential composition of 2 functions - select and then maybe insert: 

```haskell
insertBy val = do
    res <- getByValue val
    case res of
        Nothing -> Right `liftM` insert val
        Just z -> return $ Left z
```

For some databases like postgres atomicity of inserts can be achieved with expressions like `ON CONFLICT`. Actually it's already used for atomic `upsert` function. Can `insertBy` be improved similar way? Or maybe there is working alternative already?

**Comment by Vlix** [comment]
I've encountered this issue sometimes as well. We've just retried the DB action in most cases, but I think this can be fixed with something like:

```hs
insertBy val =
    withGetByValue $ (Right <$> insert val) `catch` tryGetAgain
  where
    tryGetAgain e = withGetByValue $ throwM e
    withGetByValue act = do
        res <- getByValue val
        case res of
            Nothing -> act
            Just z -> return $ Left z
```

If possible, restrict the catch to the exceptions which indicate that the insert failed because of a uniqueness constraint.

*Using `ON CONFLICT` wouldn't really help in this case, I'm afraid, since you can only `DO NOTHING` to not get an exception thrown or `DO UPDATE` to change something, which you both don't want to do on an `insertBy`. You want it to insert the value OR return the current one in the DB.*

**Comment by tim2CF** [comment]
@Vlix I ended up with introducing special boolean flag to the model called `atomicInserted` and using `upsertBy`

```haskell     res <-
       P.upsertBy
         (UniqueInvoice rh)
         this
         [InvoiceAtomicInsert P.=. False]
     let resVal = entityVal res
     pure $
       if invoiceAtomicInsert resVal
         then CreateUniqueOk res
         else
           if isGenuine this resVal
             then CreateUniqueAlreadyExists res
             else CreateUniqueReplayAttack res

```

**Comment by tysonzero** [comment]
`ON CONFLICT` + a `UNION` or a follow up query seems like it would be a big step up from the status quo.

The follow up query (and maybe the `UNION`, unclear) has a (seemingly very unlikely) race condition of another thread inserting before your insert, and then deleting before your select, giving you no row as a result.

Even with that though you end up with *much* less frequent races, and when this race condition does occur, you know with much higher certainty what happened and the DB does not throw any exceptions / close down your transaction.

**Comment by tysonzero** [comment]
https://stackoverflow.com/a/62205017

**Comment by parsonsmatt** [comment]
Yeah, I want to say I've written up an approach for this somewhere. Maybe in the esqueleto issue tracker.
